libs/ledger-live-common/src/hw/deviceAccess.ts

```diff
@@ -100,34 +101,98 @@ export const cancelDeviceAction = (transport: Transport): void => {
+  // For each device, queue of linked Promises that wait after each other, blocking any future job for a given device
+  private queuedJobsByDevice: { [deviceId: string]: QueuedJob };
+
+  private static instance: DeviceQueuedJobsManager | null = null;
+
+  // To be able to differentiate withDevice calls in our logs
+  private static jobIdCounter: number = -1;
+
+  private constructor() {
+    this.queuedJobsByDevice = {};
+  }
+
+  /**
+   * Get the singleton instance
+   */
+  static getInstance(): DeviceQueuedJobsManager {
+    if (!this.instance) {
+      this.instance = new DeviceQueuedJobsManager();
+    }
+
+    return this.instance;
+  }
+
+  /**
+   * Returns the latest queued job for a given device id
+   *
+   * @param deviceId
+   * @returns the latest QueuedJob. If none, return a queued job that can be resolved directly.
+   */
+  getLastQueuedJob(deviceId: string): QueuedJob {
+    return this.queuedJobsByDevice[deviceId] || { job: Promise.resolve(), id: -1 };
+  }
 
-// To be able to differentiate withDevice calls in our logs
-let jobId = 0;
+  /**
+   * Sets the latest queue job for a given device id
+   *
+   * Also increments the job id counter and set the newly queued job id to this new incremented value.
+   *
+   * Note: we should be fine on race conditions when updating the jobId in our use cases.
+   *
+   * @param deviceId
+   * @param jobToQueue a Promise that resolve to void, representing an async job
+   * @returns the id of the queued job
+   */
+  setLastQueuedJob(deviceId: string, jobToQueue: Promise<void>): number {
+    const id = ++DeviceQueuedJobsManager.jobIdCounter;
+    this.queuedJobsByDevice[deviceId] = { job: jobToQueue, id };
+    return id;
+  }
+}
 
 /**
  * Provides a Transport instance to a given job
  *
  * @param deviceId
  * @param options contains optional configuration
- *   - openTimeoutMs: TODO: to keep, will be used in a separate PR
+ *   - openTimeoutMs: optional timeout that limits in time the open attempt of the matching registered transport.
  */
 export const withDevice =
   (deviceId: string, options?: { openTimeoutMs?: number }) =>
   <T>(job: (t: Transport) => Observable<T>): Observable<T> =>
     new Observable(o => {
-      jobId++;
-      const tracer = new LocalTracer(LOG_TYPE, { jobId: jobId, origin: "hw:withDevice" });
-      tracer.trace(`New job for device: ${deviceId || "USB"}`);
+      const queuedJobManager = DeviceQueuedJobsManager.getInstance();
+      const previousQueuedJob = queuedJobManager.getLastQueuedJob(deviceId);
 
-      let unsubscribed;
-      let sub;
-      const deviceQueue = deviceQueues[deviceId] || { job: Promise.resolve(), id: -1 };
+      // When the new job is finished, this will unlock the associated device queue of jobs
+      let unlockQueuedJob;
+
+      const jobId = queuedJobManager.setLastQueuedJob(
+        deviceId,
+        new Promise(resolve => {
+          unlockQueuedJob = resolve;
+        }),
+      );
+
```

Comment:
could we rename `unlockQueuedJob` into `resolveQueuedJob` ?

Reply:
done ;)